src/util/stream.rs

```diff
@@ -0,0 +1,433 @@
+use std::{
+    future::Future,
+    io,
+    ops::{Deref, DerefMut},
+};
+
+use bytes::Bytes;
+use iroh::endpoint::{ReadExactError, VarInt};
+use iroh_io::AsyncStreamReader;
+use serde::{de::DeserializeOwned, Serialize};
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+
+/// An abstract `iroh::endpoint::SendStream`.
+pub trait SendStream: Send {
+    /// Send bytes to the stream. This takes a `Bytes` because iroh can directly use them.
+    fn send_bytes(&mut self, bytes: Bytes) -> impl Future<Output = io::Result<()>> + Send;
+    /// Send that sends a fixed sized buffer.
+    fn send<const L: usize>(
+        &mut self,
+        buf: &[u8; L],
+    ) -> impl Future<Output = io::Result<()>> + Send;
+    /// Sync the stream. Not needed for iroh, but needed for intermediate buffered streams such as compression.
+    fn sync(&mut self) -> impl Future<Output = io::Result<()>> + Send;
+    /// Reset the stream with the given error code.
+    fn reset(&mut self, code: VarInt) -> io::Result<()>;
+    /// Wait for the stream to be stopped, returning the error code if it was.
+    fn stopped(&mut self) -> impl Future<Output = io::Result<Option<VarInt>>> + Send;
+}
+
+/// An abstract `iroh::endpoint::RecvStream`.
+pub trait RecvStream: Send {
+    /// Receive up to `len` bytes from the stream, directly into a `Bytes`.
+    fn recv_bytes(&mut self, len: usize) -> impl Future<Output = io::Result<Bytes>> + Send;
```

Comment:
why not make it implement `tokio::io::Read` instead? and provide this as a helper method?

Reply:
but it's tricky and involves `unsafe` in some cases I think

Reply:
Yeah, I would rather have Bytes -> Bytes instead of Bytes -> complex rube goldberg machine involving Buf and unsafe -> Bytes...